File: api/src/main/java/work/beltran/discogsbrowser/api/model/SearchResults.java
package work.beltran.discogsbrowser.api.model;

import com.google.auto.value.AutoValue;
import com.google.gson.Gson;
import com.google.gson.TypeAdapter;
import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;
import java.util.List;

import rx.functions.Func1;
import work.beltran.discogsbrowser.api.model.pagination.Pagination;
import work.beltran.discogsbrowser.api.model.record.Artist;
import work.beltran.discogsbrowser.api.model.record.BasicInformation;
import work.beltran.discogsbrowser.api.model.record.Format;
import work.beltran.discogsbrowser.api.model.record.Record;

/**
 * Created by Miquel Beltran on 10.05.16.
 * More on http://beltran.work
 */
@AutoValue
public abstract class SearchResults implements RecordsWithPagination {
    public abstract Pagination getPagination();
    @SerializedName("results")
    public abstract List<SearchRecord> getSearchRecords();

    @Override
    public List<Record> getRecords() {
        return rx.Observable
                .from(getSearchRecords())
                .map(new Func1<SearchRecord, Record>() {
            @Override
            public Record call(SearchRecord searchRecord) {
                String artist = searchRecord.getTitle();
                String title = "";

                int index = searchRecord.getTitle().indexOf(" - ");
                if (index > 0) {
                    title = searchRecord.getTitle().substring(index + 3);
                    artist = searchRecord.getTitle().substring(0, index);
                }

                Artist artistObject = Artist.builder().name(artist).build();
                List<Artist> artists = new ArrayList<>();
                artists.add(artistObject);
                List<Format> formats = new ArrayList<>();
                if (searchRecord.getFormat() != null) {
                    for (String formatString : searchRecord.getFormat()) {
                        Format format = Format.builder().setName(formatString).build();
                        formats.add(format);
                    }
                }
                String year = searchRecord.getYear() != null ? searchRecord.getYear() : "";
                BasicInformation basicInformation = BasicInformation.builder()
                        .artists(artists)
                        .formats(formats)
                        .title(title)
                        .thumb(searchRecord.getThumb())
                        .year(year)
                        .build();
                return Record.builder()
                        .setInstanceId(searchRecord.getId())
                        .setBasicInformation(basicInformation)
                        .build();
            }
        }).toList().toBlocking().single();
    }

    public static TypeAdapter<SearchResults> typeAdapter(Gson gson) {
        return new AutoValue_SearchResults.GsonTypeAdapter(gson);
    }

    public static Builder builder() {
        return new AutoValue_SearchResults.Builder();
    }

    @AutoValue.Builder
    public abstract static class Builder {
        public abstract Builder setPagination(Pagination newPagination);

        public abstract Builder setSearchRecords(List<SearchRecord> newSearchRecords);

        public abstract SearchResults build();
    }
}

File: api/src/main/java/work/beltran/discogsbrowser/api/model/SearchResults.kt
package work.beltran.discogsbrowser.api.model

import com.squareup.moshi.Json
import work.beltran.discogsbrowser.api.model.pagination.Pagination
import work.beltran.discogsbrowser.api.model.record.Artist
import work.beltran.discogsbrowser.api.model.record.BasicInformation
import work.beltran.discogsbrowser.api.model.record.Format
import work.beltran.discogsbrowser.api.model.record.Record

/**
 * Created by Miquel Beltran on 10.05.16.
 * More on http://beltran.work
 */
data class SearchResults(
        override val pagination: Pagination,
        @Json(name = "results")
        val searchRecords: List<SearchRecord>
) : RecordsWithPagination {
    override val records: List<Record>
        get() = searchRecords.map {
            var artist = it.title
            var title = ""
            val index = it.title.indexOf(" - ")
            if (index > 0) {
                title = it.title.substring(index + 3)
                artist = it.title.substring(0, index)
            }
            val formats = it.format?.map { Format(it) } ?: emptyList()
            val basicInformation = BasicInformation(
                    artists = listOf(Artist(artist)),
                    formats = formats,
                    title = title,
                    thumb = it.thumb,
                    year = it.year ?: ""
            )
            Record(it.id, null, basicInformation)
        }
}
